CORE-7269 metadata add component group structure and mgm group param hash

vlajos opened the pull request:
* CORE-7269 Add component group structure to Transaction Metadata. Also calculate the number of component group from this, instead of storing that explicitly.
* CORE-7269 Store Group Parameters hash in Transaction Metadata.
* CORE-7269 Persist group Parameters when starting to use their hash in tx metadata.
* Load back stored group parameters for later verification.
* Added a kryo serializer for SignedGroupParameters.
* Added a new GroupParametersFactory create().

API: https://github.com/corda/corda-api/pull/1042

vlajos committed:
CORE-7269 Add component group structure to Transaction Metadata

vlajos committed:
CORE-7269 Add membership group parameters to Transaction Metadata

kyriathar commented on components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/factory/TransactionMetadataFactoryImpl.kt:
```diff
@@ -24,19 +25,32 @@ class TransactionMetadataFactoryImpl @Activate constructor(
         return TransactionMetadataImpl(metadata + ledgerSpecificMetadata)
     }
 
+    private fun getCurrentMgmGroupParametersHash(): String {
```
Not familiar with group parameters but I believe this should be renamed to `getCurrentMgmGroupParametersId` and maybe `InternalGroupParameters.hash` to id too, hash is an implementation detail?

vlajos replied:
I think they use `epoch` as a primary key.
`Hash` is the hash of the group parameters key-value store's binary representation what the ledger use as an identifier.
It is easy to update, but I think it reflects quite well that it is also a "fingerprint" of the the parameters.

vlajos committed:
CORE-7269 Fix CommonLedgerTest

vlajos committed:
CORE-7269 Disable serialization test

vlajos committed:
CORE-7269 make membershipgroupParametersHash to UTXO only

vlajos committed:
CORE-7269 rename some fakes to fake...

vlajos committed:
CORE-7269 fix components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoLedgerMessageProcessorTests.kt

vlajos committed:
CORE-7269 Make whole signed group parameters accessible

vlajos committed:
CORE-7269 Persist group parameters when saving their hash into a transaction

vlajos committed:
CORE-7269 Fix some tests

vlajos committed:
CORE-7269 find updates

vlajos committed:
CORE-7269 update todo comments

vlajos committed:
CORE-7269 remove unneded dependency

vlajos committed:
CORE-7269 revert unrelated change

vlajos committed:
CORE-7269 fix tests

vlajos committed:
CORE-7269 detekt

blsemo commented on components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/UtxoRepository.kt:
```diff
@@ -149,4 +150,21 @@ interface UtxoRepository {
+    fun persistSignedGroupParameters(
+        entityManager: EntityManager,
+        hash: String,
+        parameters: ByteArray,
```
I don't like the asymmetry here - your persist the serialized group paramters as `ByteArray`, but when you load it, you get back the deserialized object.

vlajos replied:
Thanks. I have updated it to use the object in persist too.

blsemo commented on components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoRepositoryImpl.kt:
```diff
@@ -441,6 +445,58 @@ class UtxoRepositoryImpl @Activate constructor(
+                    CryptoSignatureSpec((r.get(3) as String), null, null)
+                )
+            }
+            .firstOrNull()
```
Should that be `singleOrNull()`? If the hash is the primary key, than finding 2 would be definite error?

vlajos replied:
Thanks. I have updated it.

blsemo commented on testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/UtxoLedgerTest.kt:
```diff
@@ -64,11 +67,13 @@ abstract class UtxoLedgerTest : CommonLedgerTest() {
         serializationServiceNullCfg,
-        mockTransactionSignatureService(),
+        fakeTransactionSignatureService(),
         transactionMetadataFactory,
         wireTransactionFactory,
         utxoLedgerTransactionFactory,
-        mockUtxoLedgerTransactionVerificationService
+        mockUtxoLedgerTransactionVerificationService,
+        mockUtxoLedgerGroupParametersPersistenceService,
+        mockMembershipGroupReaderProvider
```
Why did you rename some of the mock services to fake services? Is there a fundamental difference between fakes and mocks here?

vlajos replied:
I think they are closer to fakes than to mocks.
Difference is not too much.
We can say naming hygiene.

blsemo reviewed (commented):
I am getting confused between `PersistSignedGroupParametersIfDoNotExistParameters`, `SignedGroupParameters`, `CordaSignedGroupParamateres` and `SignedGroupParamtersContainer` - can we simplify that somehow? It feels

vlajos committed:
CORE-7269 Avoid comparing serialized formats of objects. They may diverge. For example if the same subobject exist in multiple places. In our case there are multiple emptyList() component groups

vlajos committed:
CORE-7269 Remove SignedGroupParametersContainer. Add SignedGroupParametersKryoSerializer.

vlajos committed:
CORE-7269 Review updates. More symmetry in UtxoRepository

vlajos commented:
> I am getting confused between `PersistSignedGroupParametersIfDoNotExistParameters`, `SignedGroupParameters`, `CordaSignedGroupParamateres` and `SignedGroupParamtersContainer` - can we simplify that somehow? It feels

Yeah. It was quite confusing.
I have removed the container and added a serializer for the SignedGroupParameters to make it usable in flows.
The other parts are the standard persistence patterns.

vlajos commented on components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/consensual/tests/ConsensualLedgerMessageProcessorTests.kt:
```diff
@@ -169,16 +176,26 @@ class ConsensualLedgerMessageProcessorTests {
         assertThat(response.error).isNull()
         val entityResponse = deserializer.deserialize(response.payload.array())!!
         assertThat(entityResponse.results).hasSize(1)
-        assertThat(entityResponse.results.first()).isEqualTo(serializedTransaction)
```
It is not guaranteed that these serialized forms are the same.
It seems that having multiple empty lists in the component groups generates a different result depending on whether they are the same instances or not.

mnesbit replied:
Ah! Have we copied wrong AMQP serialization mode! In C4 there are differences between storage and messaging AMQP profiles. In particular 'storage', which is used for transactions and things that need to be consistently signed doesn't tokenise repeated object instances. It will treat them as separate items and serialize their fields all over again. Obviously in a complex object graph this can massively expand the data size, but it is only way to get a consistent serialized form.

vlajos replied:
I added some more details on slack: https://r3holdco.slack.com/archives/C03A3TQMBHR/p1681295411889109

vlajos committed:
CORE-7269 Fix todo

vlajos committed:
CORE-7269 wording

vlajos committed:
CORE-7269 simplification

vlajos committed:
CORE-7269 detekt

vlajos committed:
CORE-7269 add-back the serialization test commented out

vlajos committed:
CORE-7269 Fix dependencies

vlajos committed:
CORE-7269 Verify signed group parameters consistency

lankydan commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerGroupParametersPersistenceServiceImpl.kt:
```diff
@@ -0,0 +1,70 @@
+                FindSignedGroupParametersExternalEventFactory::class.java,
+                FindSignedGroupParametersParameters(hash.toString())
+            )
+        }.singleOrNull().let {
+            if (it == null) {
+                null
+            } else {
+                requireNotNull(it.signature){
+                    "Group parameters need to be signed."
+                }
+                requireNotNull(it.signatureSpec){
+                    "Group parameters signature need a signature specification."
+                }
+                it
+            }
+        }
+    }
```
```kotlin
        }.singleOrNull()?.also { groupParameters ->
                requireNotNull(groupParameters.signature){
                    "Group parameters need to be signed."
                }
                requireNotNull(groupParameters.signatureSpec){
                    "Group parameters signature need a signature specification."
                }
        }
```

Add the variable name because its not clear what it is, I assume its `groupParameters` but I don't actually know, hence comment.

vlajos replied:
I have updated the messages.

lankydan commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/FindSignedGroupParametersExternalEventFactory.kt:
```diff
@@ -0,0 +1,67 @@
+            requireNotNull(it.mgmSignature) {
+                "Group parameters need to be signed."
+            }
+            requireNotNull(it.mgmSignatureSpec) {
```
The avro one is `mgmSignatureSpec` but the non avro one is `signatureSpec`, do we want to align these?

vlajos replied:
It sounds a good idea. I'll open a tech debt ticket for it.

vlajos replied:
https://r3-cev.atlassian.net/browse/CORE-12457

lankydan commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/serializer/kryo/SignedGroupParametersKryoSerializer.kt:
```diff
@@ -0,0 +1,40 @@
+    property = [SandboxConstants.CORDA_UNINJECTABLE_SERVICE],
+    scope = ServiceScope.PROTOTYPE
+)
+class SignedGroupParametersKryoSerializer @Activate constructor(
```
Not saying its wrong, but why do we need a custom serializer for this class?

vlajos replied:
SignedGroupParameters have an internal lambda, which is not too Kryo friendly...
https://r3holdco.slack.com/archives/C03A3TQMBHR/p1681253250657469

lankydan commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoSignedTransactionFactoryImpl.kt:
```diff
@@ -96,13 +111,38 @@ class UtxoSignedTransactionFactoryImpl @Activate constructor(
+        val signedGroupParameters = requireNotNull(groupReader.signedGroupParameters) {
+            "Group parameters could not be accessed."
+        }
+        requireNotNull(signedGroupParameters.signature) {
+            "Group parameters need to be signed."
+        }
+        requireNotNull(signedGroupParameters.signatureSpec) {
+            "Group parameters signature need a signature specification."
+        }
```
Seen these 2 requires quite a few time through the code, should we consolidate them? Happy to push to tech debt if needed, but it will probably never be done if we do that.

vlajos replied:
I could remove two of these checks. (They are not needed since I converted back the code to use the non-avro SignedGroupParameters.)

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/factory/WireTransactionFactoryImpl.kt:
```diff
@@ -42,6 +43,12 @@ class WireTransactionFactoryImpl @Activate constructor(
     override fun create(componentGroupLists: List<List<ByteArray>>): WireTransaction {
+        val metadata =
+            parseMetadata(componentGroupLists[TransactionMetadataImpl.ALL_LEDGER_METADATA_COMPONENT_GROUP_ID].first())
+        check((metadata as TransactionMetadataInternal).getNumberOfComponentGroups() == componentGroupLists.size){
+            "Number of component groups in metadata structure description does not match with the real number!" +
+            (metadata as TransactionMetadataInternal).getNumberOfComponentGroups() + " vs " + componentGroupLists.size
+        }
```
Seen this check a number of times as well.

vlajos replied:
I removed two of these as well from the consensual/utxo signed tx factory. (The same check happened a few lines before in wire tx factory.)

lankydan commented on testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/WireTransactionExample.kt:
```diff
@@ -10,25 +10,30 @@ import net.corda.v5.application.marshalling.JsonMarshallingService
     val allGroupLists = listOf(
         listOf(canonicalJson.toByteArray()),
-    ) + componentGroups
+    ) +
```
This formatting is ugly.

vlajos replied:
I have reformatted it a bit.

lankydan commented on testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/MockMemberShipGroupReaderProvider.kt:
```diff
@@ -0,0 +1,26 @@
+    val mockGroupReader = Mockito.mock(MembershipGroupReader::class.java)
+    val mockSignedGroupParameters = Mockito.mock(SignedGroupParameters::class.java)
+
+    Mockito.`when`(mockSignedGroupParameters.bytes).thenReturn(byteArrayOf(1, 1, 11))
```
Can't we use `whenever` instead of `Mockito.when`? Or is this part of the OSGi tests? If so then ignore my comment.

vlajos replied:
If I remember right, these ones are used in OSGi tests.

vlajos committed:
CORE-7269 PR review updates

vlajos committed:
CORE-7269 Add group-policy-impl to flow-processor

lankydan commented on processors/flow-processor/build.gradle:
```diff
@@ -50,6 +50,7 @@ dependencies {
     runtimeOnly project(':libs:ledger:ledger-consensual-data')
     runtimeOnly project(':components:ledger:ledger-utxo-flow')
     runtimeOnly project(':libs:ledger:ledger-utxo-data')
+    runtimeOnly project(':libs:ledger:ledger-utxo-transaction-verifier') // todo revert with CORE-8956
```
Why are new processors needed?

vlajos replied:
Group policy provider needs membershipQueryClient and membershipPersistenceClient too.
The group policy provider is needed for the MGM public key what we use the confirm the authenticity of the signed group parameters.

vlajos replied:
(ledger-utxo-transaction-verifier was just my first guess. That one was not needed here.)

vlajos committed:
CORE-7269 Remove redundant checks

blsemo commented on components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt:
```diff
@@ -217,4 +222,24 @@ class UtxoPersistenceServiceImpl constructor(
+    }
+
+    override fun persistSignedGroupParametersIfDoNotExist(signedGroupParameters: SignedGroupParameters) {
+        val cordaSignedGroupParameters = groupParametersFactory.create(signedGroupParameters) as CordaSignedGroupParameters
+        val hash = cordaSignedGroupParameters.hash.toString()
```
Do we need to create this object just to obtain the hash? Could we make the hashing external so we don't need a roundtrip through the factory?

vlajos replied:
That sounds like bypassing the public API and depending on the implementation.
It is probably quicker, but sounds less clean for me.

vlajos replied:
I tried it. It looks better than I had expected.
I created another PR for it:
https://github.com/corda/corda-runtime-os/pull/3615
(just to keep this one green.)

vlajos committed:
CORE-7269 Populate created utxo_group_parameters.created

vlajos committed:
CORE-7269 Fix merge

charlieR3 reviewed (approved):
@corda/corda-platform-network-team  related changes LGTM.

seanbrereton reviewed (approved):
LGTM - gradle only
